src/Microsoft.Data/PrimitiveColumn.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Data
+{
+    /// <summary>
+    /// A column to hold primitive values such as int, float etc. Other value types are not really supported
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public partial class PrimitiveColumn<T> : BaseColumn
+        where T : struct
+    {
+        internal PrimitiveColumnContainer<T> _columnContainer;
+
+        internal PrimitiveColumn(string name, PrimitiveColumnContainer<T> column) : base(name, column.Length, typeof(T))
+        {
+            _columnContainer = column;
+        }
+
+        public PrimitiveColumn(string name, IEnumerable<T> values) : base(name, 0, typeof(T))
+        {
+            _columnContainer = new PrimitiveColumnContainer<T>(values);
+            Length = _columnContainer.Length;
+        }
+
+        public PrimitiveColumn(string name, bool isNullable = true) : base(name, 0, typeof(T))
+        {
+            _columnContainer = new PrimitiveColumnContainer<T>();
+        }
+
+        public override object this[long startIndex, int length] {
+            get
+            {
+                if (startIndex > Length )
+                {
+                    throw new ArgumentException($"Indexer arguments exceed Length {Length} of the Column");
+                }
+                return _columnContainer[startIndex, length];
+            }
+        }
+
+        // This method involves boxing
```

Comment:
Do we want to have `T` based indexers as well? That way we can avoid boxing in certain situations.

Reply:
First thing I tried. I don't think generic indexers are supported in C# yet. There's an issue for it here:

https://github.com/dotnet/csharplang/issues/2054

Reply:
`PrimitiveColumn<T>` is already generic. So you can return `T` here. Just like we do in `List<T>`.

https://github.com/dotnet/corefx/blob/a26fd9bb1b17510d51a9b5a2ce7aadc1a5b631ee/src/Common/src/CoreLib/System/Collections/Generic/List.cs#L145-L167

The csharplang feature you referred to is for when you want to have a generic indexer that isn't in a generic class (or the indexer returns a different type than what the class is generic for).

Reply:
Right, but I'm overriding a virtual method from `BaseColumn` though? Unless I'm not seeing something, I would have to make `BaseColumn`'s indexer generic and then deal with type conversions in the derived class? I thought boxing to return a single value seemed acceptable in this instance

Reply:
What I am thinking is adding a `new` indexer on the derived class that is typed as `T`.

At first I was thinking:

```C#
    public class BaseColumn
    {
        public virtual object this[long rowIndex]
        {
            get => throw new NotImplementedException();
            set => throw new NotImplementedException();
        }
    }

    public class PrimitiveColumn<T> : BaseColumn
       where T : struct
    {
        public override object this[long rowIndex]
        {
            get => base[rowIndex];
            set => base[rowIndex] = value;
        }

        public new T this[long rowIndex]
        {
            get => default;
            set => base[rowIndex] = value;
        }
    }
```

But that doesn't compile:

> CS0111	Type 'PrimitiveColumn<T>' already defines a member called 'this' with the same parameter types


So instead, we could do:

```C#
    public class BaseColumn
    {
        public object this[long rowIndex]
        {
            get => GetValue(rowIndex);
            set => SetValue(rowIndex, value);
        }

        protected virtual object GetValue(long rowIndex) => throw new NotImplementedException();
        protected virtual void SetValue(long rowIndex, object value) => throw new NotImplementedException();
    }

    public class PrimitiveColumn<T> : BaseColumn
       where T : struct
    {
        protected override object GetValue(long rowIndex)
        {
            return base.GetValue(rowIndex);
        }

        protected override void SetValue(long rowIndex, object value)
        {
            base.SetValue(rowIndex, value);
        }

        public new T this[long rowIndex]
        {
            get => default;
            set => base[rowIndex] = value;
        }
    }
```

This allows us to not box/unbox when the column is casted to a `PrimitiveColumn<T>`.